File: app/src/main/java/com/todoroo/andlib/sql/UnaryCriterion.java
/**
 * Copyright (c) 2012 Todoroo Inc
 *
 * <p>See the file "LICENSE" for the full license governing this code.
 */
package com.todoroo.andlib.sql;

import static com.todoroo.andlib.sql.SqlConstants.SPACE;

public class UnaryCriterion extends Criterion {

  private final Field expression;
  private final Object value;

  private UnaryCriterion(Field expression, Operator operator, Object value) {
    super(operator);
    this.expression = expression;
    this.value = value;
  }

  public static Criterion eq(Field expression, Object value) {
    return new UnaryCriterion(expression, Operator.eq, value);
  }

  /** Sanitize the given input for SQL */
  public static String sanitize(String input) {
    return input.replace("'", "''");
  }

  static Criterion neq(Field field, Object value) {
    return new UnaryCriterion(field, Operator.neq, value);
  }

  static Criterion gt(Field field, Object value) {
    return new UnaryCriterion(field, Operator.gt, value);
  }

  static Criterion lt(Field field, Object value) {
    return new UnaryCriterion(field, Operator.lt, value);
  }

  static Criterion lte(Field field, Object value) {
    return new UnaryCriterion(field, Operator.lte, value);
  }

  public static Criterion isNull(Field field) {
    return new UnaryCriterion(field, Operator.isNull, null) {
      @Override
      protected void populateOperator(StringBuilder sb) {
        sb.append(SPACE).append(operator);
      }
    };
  }

  static Criterion isNotNull(Field field) {
    return new UnaryCriterion(field, Operator.isNotNull, null) {
      @Override
      protected void populateOperator(StringBuilder sb) {
        sb.append(SPACE).append(operator);
      }
    };
  }

  public static Criterion like(Field field, String value) {
    return new UnaryCriterion(field, Operator.like, value) {
      @Override
      protected void populateOperator(StringBuilder sb) {
        sb.append(SPACE).append(operator).append(SPACE);
      }
    };
  }

  @Override
  protected void populate(StringBuilder sb) {
    beforePopulateOperator(sb);
    populateOperator(sb);
    afterPopulateOperator(sb);
  }

  @SuppressWarnings("WeakerAccess")
  void beforePopulateOperator(StringBuilder sb) {
    sb.append(expression);
  }

  @SuppressWarnings("WeakerAccess")
  void populateOperator(StringBuilder sb) {
    sb.append(operator);
  }

  @SuppressWarnings("WeakerAccess")
  void afterPopulateOperator(StringBuilder sb) {
    if (value == null) {
      return;
    }

    if (value instanceof String) {
      sb.append("'").append(sanitize((String) value)).append("'");
    } else {
      sb.append(value);
    }
  }
}

File: data/src/main/kotlin/org/tasks/data/sql/UnaryCriterion.kt
package org.tasks.data.sql

open class UnaryCriterion private constructor(private val expression: Field, operator: Operator, private val value: Any?) : Criterion(operator) {
    override fun populate() = "$expression${populateOperator()}${afterPopulateOperator()}"

    open fun populateOperator() = "$operator"

    private fun afterPopulateOperator(): Any = if (value is String) {
        "'${sanitize(value)}'"
    } else {
        value ?: ""
    }

    companion object {
        fun eq(expression: Field, value: Any?): Criterion = UnaryCriterion(expression, Operator.eq, value)

        /** Sanitize the given input for SQL  */
        fun sanitize(input: String): String = input.replace("'", "''")

        fun gt(field: Field, value: Any?): Criterion = UnaryCriterion(field, Operator.gt, value)

        fun lt(field: Field, value: Any?): Criterion = UnaryCriterion(field, Operator.lt, value)

        fun lte(field: Field, value: Any?): Criterion = UnaryCriterion(field, Operator.lte, value)

        fun isNotNull(field: Field): Criterion {
            return object : UnaryCriterion(field, Operator.isNotNull, null) {
                override fun populateOperator() = " $operator"
            }
        }

        fun isNull(field: Field): Criterion {
            return object : UnaryCriterion(field, Operator.isNull, null) {
                override fun populateOperator() = " $operator"
            }
        }

        fun like(field: Field, value: String?): Criterion {
            return object : UnaryCriterion(field, Operator.like, value) {
                override fun populateOperator() = " $operator "
            }
        }
    }
}